### PR TITLE
Bug 1232584 - [TV] FTE for usage metrics specs update for the new legal approved terms

### PR DESCRIPTION
**Pull request by Fischer-L**
This patch updates the TV FTU according to the new approved terms by legal
- Add Firefox OS Privacy Notice and Mozilla Privacy Policy
- Add links to open Firefox OS Privacy Notice and Mozilla Privacy Policy

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b8b2dea5f78d796abea666a241ada3bc4883727f

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7f168f6786b534ba0adb372cc82101a65585d123

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=6648def62e8a94c72d5fcf7cd8e1afeb6d6d8d5d

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f8e7124b1fefbf5fca1334925570e347d7525c1a

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=030e4872ca851820aebd75123459e73e55a86a47

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=288be1b0a8b512c29c8c894dc5660e14bf7a99dc

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e6de6fc4ea802b2efcf621643b2895dcb1e8922a

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ac03fffa83d5f5239328c4c15f59055fdff5f576

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=014a89f55254d44ac980b3ee3062e52bd143adc7

File: shared/js/smart-screen/simple_key_navigation.js [hunk 1]
```diff
@@ -81,26 +81,30 @@
...
   proto.focus = function skn_focus() {
-    var elem = this._List[this._focusedIndex];
-    if (elem.focus && (typeof elem.focus) === 'function') {
-      elem.focus();
-    }
-    // Fon nested case, we need to propagate child navigator event to parent.
-    if (elem instanceof SimpleKeyNavigation) {
-      elem = elem._List[elem._focusedIndex];
+    if (this._List.length > 0) {
+      var elem = this._List[this._focusedIndex];
```
Comment (Fischer-L) [inline shared/js/smart-screen/simple_key_navigation.js]: Check the length of the element list to avoid Null error caused by accessing empty list

File: tv_apps/tv-ftu/index.html [hunk 2]
```diff
@@ -27,7 +28,7 @@
     <script defer src="js/tv_ftu.js"></script>
 
     <!-- Shared L10n libraries -->
-    <script defer src="shared/js/intl/l20n.js"></script>
+    <script defer src="shared/js/l10n.js"></script>
```
Comment (Fischer-L) [inline tv_apps/tv-ftu/index.html]: Since this patch will be uplifted to v2.5 and the stable l20n.js is not in v2.5, so we use l10n.js first.

File: shared/js/smart-screen/f_t_e_wizard.js [hunk 3]
```diff
@@ -106,9 +110,11 @@
     }
 
     this._hide(this._currentPage);
-    attachTransitionEnd(
-      this._pages[++this._currentPage], this._updateNavigation.bind(this)
-    );
+    attachTransitionEnd(this._pages[++this._currentPage], function () {
+        this._updateNavigation();
+        this._ongoNext &&
```
Comment (rexboy7) [inline shared/js/smart-screen/f_t_e_wizard.js]: `onGoNext`

File: shared/js/smart-screen/simple_key_navigation.js [hunk 4]
```diff
@@ -81,26 +81,30 @@
...
 
   proto.focus = function skn_focus() {
-    var elem = this._List[this._focusedIndex];
-    if (elem.focus && (typeof elem.focus) === 'function') {
-      elem.focus();
-    }
-    // Fon nested case, we need to propagate child navigator event to parent.
-    if (elem instanceof SimpleKeyNavigation) {
-      elem = elem._List[elem._focusedIndex];
+    if (this._List.length > 0) {
```
Comment (rexboy7) [inline shared/js/smart-screen/simple_key_navigation.js]:
```
if (this._List.length < 0) {
  return
}
```

File: shared/js/smart-screen/simple_key_navigation.js [hunk 5]
```diff
@@ -81,26 +81,30 @@
...
+      }
+      this.fire('focusChanged', elem);
     }
-    this.fire('focusChanged', elem);
   };
 
   proto.blur = function skn_blur() {
-    var elem = this._List[this._focusedIndex];
-    elem.blur && (typeof elem.blur === 'function') && elem.blur();
+    if (this._List.length > 0) {
```
Comment (rexboy7) [inline shared/js/smart-screen/simple_key_navigation.js]: same above.

But actually I think keep its length non-zero is duty of its user.

File: tv_apps/tv-ftu/index.html (hunk 2 above)
Comment (rexboy7) [inline tv_apps/tv-ftu/index.html]: Please send two different patch for v2.5 and master.

File: shared/js/smart-screen/f_t_e_wizard.js (hunk 3 above)
Comment (rexboy7) [inline shared/js/smart-screen/f_t_e_wizard.js]: why do you increase the indent twice?

File: shared/js/smart-screen/f_t_e_wizard.js [hunk 6]
```diff
@@ -106,9 +110,11 @@
...
+          this._ongoNext(this._pages[this._currentPage]);
```
Comment (rexboy7) [inline shared/js/smart-screen/f_t_e_wizard.js]: There's no any need to break them into two lines.

**Comment by rexboy7** [comment]
After I pressed 'no', I can't enter the app anymore. Is that an expected behavior?

File: tv_apps/tv-ftu/js/tv_ftu.js [hunk 7]
```diff
@@ -11,35 +13,117 @@ window.addEventListener('load', function() {
...
+      this.linkNavigation = new SimpleKeyNavigation();
+
+      this.fteWizard = new FTEWizard('tvFTU.fteWizard');
+      this.fteWizard.init({
+        container: document.body,
+        pageClass: 'page',
+        buttonsClass: 'page-button',
+        propagateKeyEvent: true,
+        ongoNext: this.setupPage.bind(this),
+        ongoPrev: this.setupPage.bind(this),
```
Comment (rexboy7) [inline tv_apps/tv-ftu/js/tv_ftu.js]: I just don't understand, there is only one page on it and you need to add listener to goNext, goPrev, and init? What makes this necessary? I even think you don't have to go with `fteWizard`.

File: tv_apps/tv-ftu/style/index.css [hunk 8]
```diff
@@ -36,88 +54,111 @@ smart-button[type=rounded-corner-text] {
...
+}
+
+a#fxos-privacy-notice-link {
+  width: 37rem;
+  text-align: center;
+}
+
+a#moz-privacy-policy-link {
+  width: 31rem;
+}
```
Comment (rexboy7) [inline tv_apps/tv-ftu/style/index.css]: nit: newline

File: tv_apps/tv-ftu/index.html [hunk 9]
```diff
@@ -37,25 +38,81 @@
...
+      <section class="page-section button-section">
+        <div class="arrow-icon hidden" data-icon="arrow-down"></div>
+        <smart-button id="app-usage-page-no-button" class="page-button" type="rounded-corner-text" data-l10n-id="no"></smart-button>
+        <smart-button id="app-usage-page-yes-button" class="primary page-button page-default-button" type="rounded-corner-text" data-l10n-id="yes"></smart-button>
+      </section>
+    </article>
+
+    <article id="privacy-policy-page" class="page hidden">
+      <section class="page-section content-section" tabindex="-1">
+        <div class="page-content">
```
Comment (rexboy7) [inline tv_apps/tv-ftu/index.html]: So this is only one FTE page, but it's actually two pages? I think this is somewhat confusing since you are mixing your own paging mechanism together with FTE-wizard.

File: tv_apps/tv-ftu/js/tv_ftu.js [hunk 10]
```diff
@@ -11,35 +13,117 @@ window.addEventListener('load', function() {
...
+          this.keyNavAdapter.on('enter-keyup', this.onEnterKeyUp.bind(this));
+          this.on('focus', this.onFocus.bind(this));
+
+          this.linkNavigation.start(
+            [], SimpleKeyNavigation.DIRECTION.HORIZONTAL);
+          this.linkNavigation.on('focusChanged', (elem) => {
+            this.fire('focus', elem);
+          });
+
+          this.fteWizard._simpleKeyNavigation.on('focusChanged', (elem) => {
```
Comment (rexboy7) [inline tv_apps/tv-ftu/js/tv_ftu.js]: Don't access a private property.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f0aaa5c38b979f693e1215136f3336d3fdd4898c

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a25ac9159f163c7e558cacca1a5ea4b75894ae1b

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f138df6531aa65e24554749b659e81b516891ebf

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2cd91720b918f3cdac21f6bd658176917920966c

File: tv_apps/tv-ftu/index.html (hunk 9 above)
Comment (Fischer-L) [inline tv_apps/tv-ftu/index.html]: Remove the using of FTE-wizard, thanks

File: tv_apps/tv-ftu/js/tv_ftu.js [hunk 11]
```diff
@@ -11,35 +12,92 @@ window.addEventListener('load', function() {
...
+                             this._currentContentSection.clientHeight;
+      return unscrolledHeight;
+    },
+
+    changeNavigationTarget: function (target) {
+      if (this._currentNavigationTarget !== target) {
+
+        this._currentNavigationTarget = target;
+        this._currentArrowIcon.classList.add('hidden');
+        this.keyNav.pause();
```
Comment (rexboy7) [inline tv_apps/tv-ftu/js/tv_ftu.js]: I'm not pretty sure what this function is trying to do yet, but why not just call pause() in 'unscrolled_content'?

File: tv_apps/tv-ftu/js/tv_ftu.js [hunk 12]
```diff
@@ -11,35 +12,92 @@ window.addEventListener('load', function() {
...
+      this._currentPage.classList.remove('hidden');
+
+      this._currentNavigationTarget = null;
+      this.changeNavigationTarget('buttons');
+      if (this.getContentUnscrolledHeight() > 0) {
+        this.changeNavigationTarget('unscrolled_content');
+      }
+    },
+
+    getContentUnscrolledHeight: function () {
```
Comment (rexboy7) [inline tv_apps/tv-ftu/js/tv_ftu.js]: `Unscrolled Height` is really confusing me. I can't even find this word in dictionary. Until I read the usage below and finally realize it stands for remaining height of the scroll area. Can you find a better naming for it?

File: tv_apps/tv-ftu/js/tv_ftu.js [hunk 13]
```diff
@@ -47,26 +105,36 @@ window.addEventListener('load', function() {
...
-      switch (key) {
-        case 'up':
-          if (this._currentFocused !== this._termsSection) {
-            this.buttonNav.blur();
-            this.buttonNav.pause();
-            this._termsSection.focus();
-            this.fire('focus', this._termsSection);
-            this._arrowIcon.classList.remove('hidden');
+
+      switch (this._currentNavigationTarget) {
```
Comment (rexboy7) [inline tv_apps/tv-ftu/js/tv_ftu.js]: You have `currentNavigationTarget` and `currentFocused` and didn't have any description between `NavigationTarget` and `Focused`. I'd thought `NavigationTarget` is a kind of HTMLElement or element id at first glance but it seems `NavigationTarget` is just several types of word. I feel it pretty confused. Please try to make your variables more specific. In this case I feel `currentFocusType` or something may be more suitable.

File: tv_apps/tv-ftu/js/tv_ftu.js [hunk 14]
```diff
@@ -47,26 +105,36 @@ window.addEventListener('load', function() {
...
+        break;
+
+        case 'links':
+          // this.keyNav only handles
+          // the left/right key (horizontal key navigation)
+          // so we need extra handling of the up/down key here
+          if ((key === 'down' || key === 'right') &&
+            'moz-privacy-policy-link' === this._currentFocused.id) {
+            // If focus on the last link, moz-privacy-policy-link,
+            // then change navigation target to buttons
```
Comment (rexboy7) [inline tv_apps/tv-ftu/js/tv_ftu.js]: What will happen if links are located vertically with each other after localized?

Comment (rexboy7) [inline tv_apps/tv-ftu/js/tv_ftu.js]: If the case above can't be handled correctly, we may need to switch to `spatialNavigator` rather than just use `simpleNavigator`.

Comment (rexboy7) [inline tv_apps/tv-ftu/js/tv_ftu.js]: After some discussion I realize this decision has been discussed with UX, then I'm ok with it.

File: tv_apps/tv-ftu/js/tv_ftu.js [hunk 15]
```diff
@@ -86,6 +154,36 @@ window.addEventListener('load', function() {
...
+        case 'moz-privacy-policy-link':
+        case 'fxos-privacy-notice-link':
+          var privacySection =
+            document.querySelector('#privacy-policy-page > .content-section');
+          var mozPrivacy =
+            document.querySelector('#moz-privacy-policy-content');
+          var fxosPrivacy =
+            document.querySelector('#fxos-privacy-notice-content');
+
+          if (this._currentFocused.id === 'moz-privacy-policy-link') {
```
Comment (rexboy7) [inline tv_apps/tv-ftu/js/tv_ftu.js]: Do you consider to make these two in different pages?

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d9177aad28c7adc38bba5e58f4c35ea75d62c1f4
